File: nebras-app/app/src/main/kotlin/com/nebras/mobile/core/error/ErrorHandler.kt
package com.nebras.mobile.core.error

import okhttp3.Response
import org.json.JSONObject
import java.io.IOException
import java.io.InterruptedIOException
import java.net.ConnectException
import java.net.SocketTimeoutException
import java.net.UnknownHostException
import javax.net.ssl.SSLException

/**
 * يحوّل استثناءات طبقة الشبكة إلى [Failure] مفهومة في طبقة الـ domain.
 * هذا **المكان الوحيد** الذي تُعالَج فيه استثناءات OkHttp — الـ domain لا
 * يرى استثناء شبكة قطّ (مقابل `error_handler.dart` مع Dio).
 */
object ErrorHandler {

    fun handleException(exception: Throwable): Failure = when (exception) {
        is SocketTimeoutException, is InterruptedIOException ->
            TimeoutFailure("Connection timeout. Please check your internet connection.")

        is UnknownHostException, is ConnectException ->
            NetworkFailure("Network error. Please check your internet connection.")

        is SSLException -> NetworkFailure("SSL certificate error")

        is IOException -> NetworkFailure(
            exception.message ?: "Network error. Please check your internet connection.",
        )

        else -> UnknownFailure(exception.message ?: "An unexpected error occurred")
    }

    /** يترجم ردّاً غير ناجح إلى [Failure] بحسب رمز الحالة. */
    fun handleBadResponse(statusCode: Int, body: String?): Failure {
        val message = extractErrorMessage(statusCode, body)

        if (statusCode in 400..499) {
            return when (statusCode) {
                401 -> UnauthorizedFailure(message)
                403 -> ForbiddenFailure(message)
                404 -> NotFoundFailure(message)
                else -> ClientFailure(message, statusCode)
            }
        }
        if (statusCode >= 500) return ServerFailure(message, statusCode)
        return UnknownFailure(message)
    }

    fun handleResponse(response: Response, body: String?): Failure =
        handleBadResponse(response.code, body)

    /** يستخرج رسالة الخطأ من صيغ ردود الـ API الشائعة. */
    private fun extractErrorMessage(statusCode: Int, body: String?): String {
        if (body.isNullOrBlank()) return "Error $statusCode"
        return runCatching {
            val json = JSONObject(body)
            json.optString("message").takeIf { it.isNotEmpty() }
                ?: json.optString("error").takeIf { it.isNotEmpty() }
                ?: json.optString("detail").takeIf { it.isNotEmpty() }
                ?: "Error $statusCode"
        }.getOrDefault("Error $statusCode")
    }
}

File: app-flutter/lib/core/error/error_handler.dart
import 'package:dio/dio.dart';
import 'package:nebras_mobile_app/core/error/failures.dart';

/// Error Handler
/// Translates Dio exceptions into domain-friendly Failure objects
/// This is the ONLY place where Dio exceptions are handled
/// Domain layer never sees Dio exceptions
class ErrorHandler {
  ErrorHandler._();

  /// Convert DioException to Failure
  static Failure handleDioException(DioException exception) {
    switch (exception.type) {
      case DioExceptionType.connectionTimeout:
      case DioExceptionType.sendTimeout:
      case DioExceptionType.receiveTimeout:
        return TimeoutFailure(
          'Connection timeout. Please check your internet connection.',
        );

      case DioExceptionType.connectionError:
        return const NetworkFailure(
          'Network error. Please check your internet connection.',
        );

      case DioExceptionType.badResponse:
        return _handleBadResponse(exception);

      case DioExceptionType.cancel:
        return const CancellationFailure('Request was cancelled');

      case DioExceptionType.badCertificate:
        return const NetworkFailure('SSL certificate error');

      case DioExceptionType.unknown:
        return UnknownFailure(
          exception.message ?? 'An unexpected error occurred',
        );
    }
  }

  /// Handle bad response (status code errors)
  static Failure _handleBadResponse(DioException exception) {
    final statusCode = exception.response?.statusCode;
    final message = _extractErrorMessage(exception.response);

    if (statusCode == null) {
      return UnknownFailure(message);
    }

    // Client errors (4xx)
    if (statusCode >= 400 && statusCode < 500) {
      switch (statusCode) {
        case 401:
          return UnauthorizedFailure(message);
        case 403:
          return ForbiddenFailure(message);
        case 404:
          return NotFoundFailure(message);
        default:
          return ClientFailure(message, statusCode);
      }
    }

    // Server errors (5xx)
    if (statusCode >= 500) {
      return ServerFailure(message, statusCode);
    }

    return UnknownFailure(message);
  }

  /// Extract error message from response
  static String _extractErrorMessage(Response? response) {
    if (response == null) {
      return 'An error occurred';
    }

    try {
      final data = response.data;
      
      // Try to extract message from common API response formats
      if (data is Map<String, dynamic>) {
        return data['message'] ??
            data['error'] ??
            data['detail'] ??
            'Error ${response.statusCode}';
      }

      return 'Error ${response.statusCode}';
    } catch (e) {
      return 'Error ${response.statusCode}';
    }
  }

  /// Handle generic exceptions
  static Failure handleException(Object exception) {
    if (exception is DioException) {
      return handleDioException(exception);
    }

    return UnknownFailure(exception.toString());
  }
}
